### PR TITLE
Shadow DOM support: widgets task + tests

### PR DESCRIPTION
**Pull request by AlexKamaev**

### CODE REVIEW

File: js/core/utils/shadow_dom.js [hunk 1]
```diff
@@ -60,3 +60,34 @@ export function addShadowDomStyles($element) {
...
+
+    return rect && x >= rect.left && x < rect.right && y >= rect.top && y < rect.bottom;
+}
+
+export function getShadowElementsFromPoint(x, y, root) {
+    const result = [];
+    const elementQueue = [root];
+
+    while(elementQueue.length) {
+        const el = elementQueue.shift();
```
Comment (ksercs) [inline js/core/utils/shadow_dom.js:99]: This code gives us `O(n^2)` asymptotic. If we need to implement BFS it's good idea to implement queue data structure (using Object foe example)

File: js/core/utils/shadow_dom.js [hunk 2]
```diff
@@ -60,3 +60,34 @@ export function addShadowDomStyles($element) {
...
+
+        for(let i = 0; i < el.childNodes.length; i++) {
+            const childNode = el.childNodes[i];
+
+            // eslint-disable-next-line no-undef
+            if(childNode.nodeType === Node.ELEMENT_NODE &&
+               isPositionInElementRectangle(childNode, x, y) &&
+               // eslint-disable-next-line no-undef
+               getComputedStyle(childNode).pointerEvents !== 'none'
```
Comment (ksercs) [inline js/core/utils/shadow_dom.js:108]: This code will trigger relayout, which is very expensive operation. Is it really necessary for our use cases?

File: js/ui/draggable.js [hunk 3]
```diff
@@ -866,7 +866,7 @@ const Draggable = DOMComponent.inherit({
         const $targetDraggableElement = this.$element();
 
         const mousePosition = getMousePosition(e);
-        const elements = domAdapter.elementsFromPoint(mousePosition.x, mousePosition.y);
+        const elements = domAdapter.elementsFromPoint(mousePosition.x, mousePosition.y, this.element());
```
Comment (ksercs) [inline js/ui/draggable.js:869]: Does this work? Seem like it gonna be
```suggestion
        const elements = domAdapter.elementsFromPoint(mousePosition.x, mousePosition.y, this.$element().get(0));
```

File: testing/tests/DevExpress.ui.widgets/deferRendering.animation.tests.js [hunk 4]
```diff
@@ -23,7 +16,16 @@ QUnit.testStart(function() {
...
+    const style = document.createElement('style');
+    style.innerHTML =
+        '.test-staggering-item,\
+        test-no-staggering-item {\
+            height: 10px;\
+            width: 10px;\
+        }';
+
     $('#qunit-fixture').html(markup);
+    document.body.appendChild(style);
```
Comment (ksercs) [inline testing/tests/DevExpress.ui.widgets/deferRendering.animation.tests.js:29]: why is this change necessary?

File: testing/tests/DevExpress.ui.widgets/diagramParts/toolbox.tests.js [hunk 5]
```diff
@@ -27,24 +27,24 @@ QUnit.module('Toolbox', {
 }, () => {
     test('should render if toolbox.visibility is "visible"', function(assert) {
         this.instance.option('toolbox.visibility', 'visible');
-        const $accordion = $('body').find(Consts.TOOLBOX_ACCORDION_SELECTOR);
+        const $accordion = $('#qunit-fixture').find(Consts.TOOLBOX_ACCORDION_SELECTOR);
```
Comment (ksercs) [inline testing/tests/DevExpress.ui.widgets/diagramParts/toolbox.tests.js:30]: why cannot we use body here?

File: testing/tests/DevExpress.ui.widgets/overlay.tests.js [hunk 6]
```diff
@@ -127,6 +128,8 @@ const toSelector = (cssClass) => `.${cssClass}`;
 
 const moduleConfig = {
     beforeEach: function() {
+        viewPort($('#qunit-fixture').addClass(VIEWPORT_CLASS));
```
Comment (ksercs) [inline testing/tests/DevExpress.ui.widgets/overlay.tests.js]: shouldn't we restore viewport value in afterEach?

File: testing/tests/DevExpress.ui.widgets/tabPanel.width.tests.js [hunk 7]
```diff
@@ -15,6 +16,8 @@ QUnit.module('Tabs width', () => {
             this.$tabPanel = $('<div>');
             this.assert = assert;
             this.setWidthApproach = setWidthApproach;
+
+            addShadowDomStyles($('#qunit-fixture'));
```
Comment (ksercs) [inline testing/tests/DevExpress.ui.widgets/tabPanel.width.tests.js]: why should we do this? I thought that shadow dom styles are added on component render

File: js/core/utils/shadow_dom.js (hunk 1 above)
Comment (AlexKamaev) [reply js/core/utils/shadow_dom.js:99]: ok

File: js/core/utils/shadow_dom.js (hunk 2 above)
Comment (AlexKamaev) [reply js/core/utils/shadow_dom.js:108]: unfortunately we need to check this for overlays.
I am not sure if it's needed only for overlays, but if so, we can skip just overlay elements

File: js/ui/draggable.js (hunk 3 above)
Comment (AlexKamaev) [reply js/ui/draggable.js:869]: both these approaches work

File: testing/tests/DevExpress.ui.widgets/deferRendering.animation.tests.js (hunk 4 above)
Comment (AlexKamaev) [reply testing/tests/DevExpress.ui.widgets/deferRendering.animation.tests.js:29]: we need to style for the `.test-staggering-item` element, which is in root document, not in shadow root, so we need to attach styles to the body element

File: testing/tests/DevExpress.ui.widgets/diagramParts/toolbox.tests.js (hunk 5 above)
Comment (AlexKamaev) [reply testing/tests/DevExpress.ui.widgets/diagramParts/toolbox.tests.js:30]: because we do not have the `body` element inside the shadow root

File: testing/tests/DevExpress.ui.widgets/overlay.tests.js (hunk 6 above)
Comment (AlexKamaev) [reply testing/tests/DevExpress.ui.widgets/overlay.tests.js]: well. It was not being restored in the previous version. However, using the `beforeEach` is redundant. I moved this code to QUnit.testStart. Thank you for pointing me to this

File: testing/tests/DevExpress.ui.widgets/tabPanel.width.tests.js (hunk 7 above)
Comment (AlexKamaev) [reply testing/tests/DevExpress.ui.widgets/tabPanel.width.tests.js]: I'll check it. Probably it's a fix that is not actual anymore

File: js/core/utils/shadow_dom.js [hunk 8]
```diff
@@ -60,3 +60,55 @@ export function addShadowDomStyles($element) {
...
+            ) {
+                elementQueue.push(childNode);
+                result.push(childNode);
+            }
+        }
+    }
+
+    return result.reverse();
```
Comment (ksercs) [inline js/core/utils/shadow_dom.js]: seems like we don't need `result` array anymore - we can just get `items` array from our queue

File: testing/tests/DevExpress.ui.widgets/deferRendering.animation.tests.js (hunk 4 above)
Comment (ksercs) [reply testing/tests/DevExpress.ui.widgets/deferRendering.animation.tests.js:29]: i think it's good idea to add a comment for such specific case
`// NOTE: staggering item is placed in body outside of shadow root

File: testing/tests/DevExpress.ui.widgets/tabPanel.width.tests.js (hunk 7 above)
Comment (AlexKamaev) [reply testing/tests/DevExpress.ui.widgets/tabPanel.width.tests.js]: not needed, thank you for pointing out

File: testing/tests/DevExpress.ui.widgets/deferRendering.animation.tests.js (hunk 4 above)
Comment (AlexKamaev) [reply testing/tests/DevExpress.ui.widgets/deferRendering.animation.tests.js:29]: ok

File: js/core/utils/shadow_dom.js (hunk 8 above)
Comment (AlexKamaev) [reply js/core/utils/shadow_dom.js]: we can use items, but we still need to reverse them and exclude the shadow root element

Comment (ksercs) [reply js/core/utils/shadow_dom.js]: My idea is that we may not add an additional memory. After your last change we still have an additional array, but also make additional linear iteration.

Smth like this:
```javascript
const result = queue.getItems();
result.pop();

return result;
```

**Review by ksercs — APPROVED** [review]

**Review by Zedwag — APPROVED** [review]
